src/backend/distributed/planner/merge_planner.c

```diff
@@ -156,6 +157,32 @@ CreateMergePlan(uint64 planId, Query *originalQuery, Query *query,
 }
 
 
+/*
+ * GetMergeJoinConditionList returns all the Join conditions from the ON clause
+ */
+List *
+GetMergeJoinConditionList(Query *mergeQuery)
+{
+#if PG_VERSION_NUM >= PG_VERSION_17
+	List *mergeJoinConditionList = NIL;
+	if (IsA(mergeQuery->mergeJoinCondition, List))
+	{
+		mergeJoinConditionList = (List *) mergeQuery->mergeJoinCondition;
+	}
+	else
+	{
+		Node *joinClause =
+			eval_const_expressions(NULL, mergeQuery->mergeJoinCondition);
+		joinClause = (Node *) canonicalize_qual((Expr *) joinClause, false);
+		mergeJoinConditionList = make_ands_implicit((Expr *) joinClause);
+	}
+#else
+	List *mergeJoinConditionList = WhereClauseList(mergeQuery->jointree);
+#endif
+	return mergeJoinConditionList;
+}
+
```

Comment:
@tejeswarm `GetMergeJoinConditionList()` looks sound, but below is an alternative that avoids reproducing the code from `WhereClauseList()` and I believe achieves the same effect. It is based on the assumption that the fromlist of a merge query jointree only contains range table references (and, from observation, a range table reference to the source relation). Range table references are not processed by the WhereClauseList() walker; they are treated as leaf nodes, and therefore a nop, by Postgres' `expression_tree_walker()`. So we can make a temporary From expression with an empty fromlist and let `WhereClauseList()` do the work on the query's `mergeJoinCondition`

```
List *
GetMergeJoinConditionList(Query *mergeQuery)
{
	FromExpr *merge_jointree = NULL;
#if PG_VERSION_NUM >= PG_VERSION_17
    /*
	 * In Postgres 17, the query tree has a specific field for the merge condition.
	 * For deriving the WhereClauseList from the merge condition, we construct a dummy
	 * jointree with an empty fromlist. This works because the fromlist of a merge query
	 * join tree consists of range table references only, and range table references are
	 * disregarded by the WhereClauseList() walker.
	 */
	merge_jointree = makeFromExpr(NIL, mergeQuery->mergeJoinCondition);
#else
	merge_jointree = mergeQuery->jointree;
#endif

	List *mergeJoinConditionList = WhereClauseList(merge_jointree);
	return mergeJoinConditionList;
}
```

Reply:
Nice idea! I think avoids unnecessary code. Done in the next commit